numba_dpex/core/kernel_interface/func.py

```diff
@@ -6,100 +6,224 @@
-        target_context=dpex_target.target_context,
-        typing_context=dpex_target.typing_context,
-        args=args,
-        is_kernel=False,
-        debug=debug,
-    )
-    func = cres.library.get_function(cres.fndesc.llvm_func_name)
-    cres.target_context.mark_ocl_device(func)
-    devfn = DpexFunction(cres)
+class DpexFunction(object):
+    """Class to materialize dpex function
 
-    class _function_template(ConcreteTemplate):
-        key = devfn
-        cases = [cres.signature]
+    Helper class to eager compile a specialized numba_dpex.func
+    decorated Python function into a LLVM function with spir_func
+    calling convention.
 
-    cres.typing_context.insert_user_function(devfn, _function_template)
-    libs = [cres.library]
-    cres.target_context.insert_user_function(devfn, cres.fndesc, libs)
-    return devfn
+    A specialized numba_dpex.func decorated Python function is one
+    where the user has specified a signature or a list of signatures
+    for the function. The function gets compiled as soon as the Python
+    program is loaded, i.e., eagerly, instead of JIT compilation once
+    the function is invoked.
+    """
 
+    def __init__(self, pyfunc, debug=None):
+        """Constructor for DpexFunction
 
-def compile_func_template(pyfunc, debug=None):
-    """Compile a DpexFunctionTemplate"""
+        Args:
+            pyfunc (function): A python function to be compiled.
+            debug (object, optional): Debug option for compilation.
+                Defaults to None.
+        """
+        self._pyfunc = pyfunc
+        self._debug = debug
 
-    dft = DpexFunctionTemplate(pyfunc, debug=debug)
+    def compile(self, arg_types, return_types):
+        """The actual compilation function.
 
-    class _function_template(AbstractTemplate):
-        key = dft
+        Args:
+            arg_types (tuple): Function argument types in a tuple.
+            return_types (numba.core.types.scalars.Integer):
+                An integer value to specify the return type.
 
-        def generic(self, args, kws):
-            if kws:
-                raise AssertionError("No keyword arguments allowed.")
-            return dft.compile(args)
+        Returns:
+            numba.core.compiler.CompileResult: The compiled result
+        """
 
-    dpex_target.typing_context.insert_user_function(dft, _function_template)
-    return dft
+        cres = compile_with_dpex(
+            pyfunc=self._pyfunc,
+            pyfunc_name=self._pyfunc.__name__,
+            return_type=return_types,
+            target_context=dpex_target.target_context,
+            typing_context=dpex_target.typing_context,
+            args=arg_types,
+            is_kernel=False,
+            debug=self._debug,
+        )
+        func = cres.library.get_function(cres.fndesc.llvm_func_name)
+        cres.target_context.mark_ocl_device(func)
+
+        return cres
 
 
 class DpexFunctionTemplate(object):
     """Unmaterialized dpex function"""
```

Comment:
Docstring should be updated to:

"Helper class to JIT compile an unspecialized numba_dpex.func
 decorated Python function into an LLVM function with spir_func calling convention."